# Теория вероятностей и математическая статистика
## Урок 2. Дискретные распределения вероятностей

In [1]:
import numpy as np
from math import factorial

In [2]:
def combinations(n, k):
    return np.math.factorial(n) / (np.math.factorial(k) * np.math.factorial(n - k))

In [3]:
def bernoulli_distribution(n, k, p, q):
    return combinations(n, k) * p**k * q**(n-k)

In [4]:
def poisson_distribution(m, l):
    return (l**m / np.math.factorial(m)) * 2.72**(-l)

**Задача 1**  
Вероятность того, что стрелок попадет в мишень, выстрелив один раз, равна 0.8. Стрелок выстрелил 100 раз. Найдите вероятность того, что стрелок попадет в цель ровно 85 раз.

In [5]:
p = 0.8 # Вероятность того, что стрелок попадет в мишень, выстрелив один раз
q = 0.2 # Вероятность того, что стрелок НЕ попадет в мишень, выстрелив один раз
n = 100 # раз стрелок выстрелил
k = 85 # раз стрелок попал в цель

round(bernoulli_distribution(n, k, p, q), 3)

0.048

----------------------------------------

**Ответ:** P=0.048

-------------------------------------------

**Задача 2**  
Вероятность того, что лампочка перегорит в течение первого дня эксплуатации, равна 0.0004. В жилом комплексе после ремонта в один день включили 5000 новых лампочек. Какова вероятность, что ни одна из них не перегорит в первый день? Какова вероятность, что перегорят ровно две?

In [6]:
p = 0.0004 # Вероятность того, что лампочка перегорит в течение первого дня эксплуатации
n = 5000 # лампочек
l = p * n # среднее количество наступления события
m1 = 0 # раз лампочки перегорят
m2 = 2 # раза лампочки перегорят

no_burnout = poisson_distribution(m1, l)
two_burnout = poisson_distribution(m2, l)

print(f'Ни одна лампочка не перегорит: {round(no_burnout, 3)}')
print(f'2 лампочки перегорят: {round(two_burnout, 3)}')

Ни одна лампочка не перегорит: 0.135
2 лампочки перегорят: 0.27


--------------------------------

**Ответ:**  
1) P(ни одна лампочка не перегорит) = 0.135  
2) P(2 лампочки перегорят) = 0.27

**Задача 3:**  
Монету подбросили 144 раза. Какова вероятность, что орел выпадет ровно 70 раз?

In [7]:
n = 144 # раза подбросили монетку
k = 70 # раз выпал орел
p = 0.5 # вероятность того, что выпадет орел
q = 0.5 # вероятность того,что орел НЕ выпадет, то есть выпадет решка

round(bernoulli_distribution(n, k, p, q), 3)

0.063

------------------------

**Ответ:** P = 0.063

-------------------------

**Задача 4:**  
В первом ящике находится 10 мячей, из которых 7 - белые. Во втором ящике - 11 мячей, из которых 9 белых. Из каждого ящика вытаскивают случайным образом по два мяча. Какова вероятность того, что все мячи белые? Какова вероятность того, что ровно два мяча белые? Какова вероятность того, что хотя бы один мяч белый?

In [8]:
# Какова вероятность того, что все мячи белые?
all_white = (7/10 * 6/9) * (9/11 * 8/10)

# Какова вероятность того, что ровно два мяча белые? 
#          (ББ)(ЧЧ)       либо     (ЧЧ)(ББ)      либо     (БЧ)(БЧ)      либо     (ЧБ)(ЧБ)      либо    (БЧ)(ЧБ)      либо (ЧБ)(БЧ)
two_white = (7/10*6/9)*(2/11*1/10)+(3/10*2/9)*(9/11*8/10)+(7/10*3/9)*(9/11*2/10)+(3/10*7/9)*(2/11*9/10)+(7/10*3/9)*(2/11*9/10)+(3/10*7/9)*(9/11*2/10)

# Какова вероятность того, что хотя бы один мяч белый?
no_white = (3/10 * 2/9) * (2/11 * 1/10) # вероятность того, что не будет ни одного белого мяча
at_least_one = 1 - no_white

print(f'Все мячи белые: {round(all_white, 3)}')
print(f'Ровно 2 мяча белых: {round(two_white, 3)}')
print(f'Хотя бы 1 белый мяч: {round(at_least_one, 3)}')

Все мячи белые: 0.305
Ровно 2 мяча белых: 0.205
Хотя бы 1 белый мяч: 0.999


---------------

**Ответ:**  
1) P(все мячи белые) = 0.305    
2) P(ровно 2 мяча белых) = 0.205    
3) P(хотя бы 1 белый мяч) = 0.999  

----------------------------------------